File: src/main/java/org/processmining/models/base_implementation/BaseTransition.java
package org.processmining.models.base_implementation;

import org.deckfour.xes.model.XTrace;
import org.processmining.models.GenerationDescription;
import org.processmining.models.MovementResult;
import org.processmining.models.abstract_net_representation.AbstractPetriNode;
import org.processmining.models.abstract_net_representation.Place;
import org.processmining.models.abstract_net_representation.Token;
import org.processmining.models.abstract_net_representation.Transition;
import org.processmining.utils.LoggingSingleton;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Ivan Shugurov on 31.10.2014.
 */
public class BaseTransition extends Transition<Place<Token>>
{
    protected BaseTransition(org.processmining.models.graphbased.directed.petrinet.elements.Transition transition, GenerationDescription generationDescription, Place<Token>[] inputPlaces, Place<Token>[] outputPlaces)
    {
        super(transition, generationDescription, inputPlaces, outputPlaces);
    }

    @Override
    public MovementResult move(XTrace trace)
    {
        consumeTokens();
        LoggingSingleton.log(trace, getNode());
        produceTokens();
        return new MovementResult();
    }

    protected void produceTokens()
    {
        for (AbstractPetriNode node : getOutputPlaces())
        {
            Place place = (Place) node;
            Token token = new Token();
            place.addToken(token);
        }
    }

    protected void consumeTokens()
    {
        for (Place place : getInputPlaces())
        {
            place.consumeToken();
        }
    }

    public static class BaseTransitionBuilder
    {
        private final org.processmining.models.graphbased.directed.petrinet.elements.Transition transition;
        private final GenerationDescription description;
        private final List<Place> inputPlaces = new ArrayList<Place>();
        private final List<Place> outputPlaces = new ArrayList<Place>();

        public BaseTransitionBuilder(org.processmining.models.graphbased.directed.petrinet.elements.Transition transition, GenerationDescription/*TODO ужен свой тип?*/ description)
        {
            if (transition == null)
            {
                throw new IllegalArgumentException("Transition cannot be equal to null");
            }

            if (description == null)
            {
                throw new IllegalArgumentException("Generation description cannot be equal to null");
            }

            this.transition = transition;
            this.description = description;
        }

        public BaseTransitionBuilder inputPlace(Place<Token> inputPlace)
        {
            if (inputPlace == null)
            {
                throw new NullPointerException("Place cannot be equal to null");
            }

            inputPlaces.add(inputPlace);

            return this;
        }

        public BaseTransitionBuilder outputPlace(Place<Token> outputPlace)
        {
            if (outputPlace == null)
            {
                throw new NullPointerException("Output place cannot be null");
            }

            outputPlaces.add(outputPlace);

            return this;
        }

        public BaseTransition build()
        {
            @SuppressWarnings("unchecked")
            Place<Token>[] inputPlacesArray = inputPlaces.toArray(new Place[inputPlaces.size()]);

            @SuppressWarnings("unchecked")
            Place<Token>[] outputPlacesArray = outputPlaces.toArray(new Place[outputPlaces.size()]);

            return new BaseTransition(transition, description, inputPlacesArray, outputPlacesArray);
        }
    }
}

File: src/main/java/org/processmining/models/base_implementation/BaseTransition.kt
package org.processmining.models.base_implementation

import org.deckfour.xes.model.XTrace
import org.processmining.models.GenerationDescription
import org.processmining.models.MovementResult
import org.processmining.models.abstract_net_representation.Place
import org.processmining.models.abstract_net_representation.Token
import org.processmining.models.abstract_net_representation.Transition
import org.processmining.utils.LoggingSingleton

import java.util.ArrayList

/**
 * Created by Ivan Shugurov on 31.10.2014.
 */
// TODO сюда добавить inhibitor and reset arcs
open class BaseTransition protected constructor(
        transition: org.processmining.models.graphbased.directed.petrinet.elements.Transition,
        generationDescription: GenerationDescription,
        inputPlaces: List<Place<Token>>,
        outputPlaces: List<Place<Token>>,
        inputInhibitorArcPlaces: List<Place<Token>> = listOf(),
        inputResetArcPlaces: List<Place<Token>> = listOf()
) : Transition<Place<Token>>(transition, generationDescription, inputPlaces, outputPlaces, inputInhibitorArcPlaces, inputResetArcPlaces) {
    
    override fun move(trace: XTrace): MovementResult<*>? {
        consumeTokens()
        LoggingSingleton.log(trace, node)
        produceTokens()
        return MovementResult<Token>()
    }
    
    protected fun produceTokens() {
        outputPlaces.forEach { place -> place.addToken(Token()) }
    }
    
    protected fun consumeTokens() {
        inputPlaces.forEach { place -> place.consumeToken() }
        // inputInhibitorArcPlaces have no tokens
        inputResetArcPlaces.forEach { place -> while(place.hasTokens()) place.consumeToken() }
    }
    
    companion object {
        
        fun build(transition: org.processmining.models.graphbased.directed.petrinet.elements.Transition,
                  description: GenerationDescription, init: BaseTransitionBuilder.() -> Unit): BaseTransition {
            return BaseTransitionBuilder(transition, description).apply { init() }.build()
        }
        
        // TODO pavel: remove builder at all
        class BaseTransitionBuilder(
                private val transition: org.processmining.models.graphbased.directed.petrinet.elements.Transition,
                private val description: GenerationDescription/*TODO ужен свой тип?*/
        ) {
            private val inputPlaces = ArrayList<Place<Token>>()
            private val outputPlaces = ArrayList<Place<Token>>()
            
            fun inputPlace(inputPlace: Place<Token>): BaseTransitionBuilder {
                inputPlaces.add(inputPlace)
                return this
            }
            
            fun outputPlace(outputPlace: Place<Token>): BaseTransitionBuilder {
                outputPlaces.add(outputPlace)
                return this
            }
            
            fun build(): BaseTransition {
                val inputPlacesArray = inputPlaces
                val outputPlacesArray = outputPlaces
                
                return BaseTransition(transition, description, inputPlacesArray, outputPlacesArray)
            }
        }
    }
}
